vpr/src/pack/verify_flat_placement.cpp

```diff
@@ -0,0 +1,94 @@
+        if (flat_placement_info.blk_x_pos[blk_id] == FlatPlacementInfo::UNDEFINED_POS ||
+            flat_placement_info.blk_y_pos[blk_id] == FlatPlacementInfo::UNDEFINED_POS) {
+            VTR_LOG_ERROR(
+                "Atom block %s has an undefined position in the flat placement.\n",
+                atom_netlist.block_name(blk_id).c_str());
+            num_errors++;
+        }
+    }
+
+    // 2. Verify that every atom block has non-negative position values.
+    //
+    // Since the device may not be sized yet, we cannot check if the positions
+    // are within the bounds of the device, but if any position value is
+    // negative (and is not undefined) we know that it is invalid.
+    for (AtomBlockId blk_id : atom_netlist.blocks()) {
+        float blk_x_pos = flat_placement_info.blk_x_pos[blk_id];
+        float blk_y_pos = flat_placement_info.blk_y_pos[blk_id];
+        int blk_sub_tile = flat_placement_info.blk_sub_tile[blk_id];
+        int blk_site_idx = flat_placement_info.blk_site_idx[blk_id];
+        if ((blk_x_pos < 0.f && blk_x_pos != FlatPlacementInfo::UNDEFINED_POS) ||
+            (blk_y_pos < 0.f && blk_y_pos != FlatPlacementInfo::UNDEFINED_POS) ||
+            (blk_sub_tile < 0 && blk_sub_tile != FlatPlacementInfo::UNDEFINED_SUB_TILE) ||
+            (blk_site_idx < 0 && blk_site_idx != FlatPlacementInfo::UNDEFINED_SITE_IDX)) {
+            VTR_LOG_ERROR(
+                "Atom block %s is placed at an invalid position on the FPGA.\n",
+                atom_netlist.block_name(blk_id).c_str());
+            num_errors++;
+        }
+    }
+
+    // 3. Verify that every atom in each molecule has the same position.
+    //
+    // TODO: In the future, we can support if some of the atoms are undefined,
+    //       but that can be fixed-up before calling this method.
+    std::vector<t_pack_molecule*> molecules = prepacker.get_molecules_vector();
+    for (t_pack_molecule* mol : molecules) {
+        AtomBlockId root_blk_id = mol->atom_block_ids[mol->root];
+        float root_pos_x = flat_placement_info.blk_x_pos[root_blk_id];
+        float root_pos_y = flat_placement_info.blk_y_pos[root_blk_id];
+        int root_sub_tile = flat_placement_info.blk_sub_tile[root_blk_id];
+        for (AtomBlockId mol_blk_id : mol->atom_block_ids) {
+            if (!mol_blk_id.is_valid())
+                continue;
+            if (flat_placement_info.blk_x_pos[mol_blk_id] != root_pos_x ||
+                flat_placement_info.blk_y_pos[mol_blk_id] != root_pos_y ||
+                flat_placement_info.blk_sub_tile[mol_blk_id] != root_sub_tile) {
+                VTR_LOG_ERROR(
+                    "Molecule with root atom block %s contains atom block %s "
+                    "which is not at the same position as the root atom "
+                    "block.\n",
+                    atom_netlist.block_name(root_blk_id).c_str(),
+                    atom_netlist.block_name(mol_blk_id).c_str());
+                num_errors++;
```

Comment:
I'd test with large carry chains to see what molecule is created -- does it span multiple logic blocks?  If so, the x and y of all blocks in the molecule wouldn't have to be the same (and in fact shouldn't be).